staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go

```diff
@@ -513,6 +454,55 @@ func (cnc *CloudNodeController) getNodeModifiersFromCloudProvider(ctx context.Co
 	return nodeModifiers, nil
 }
 
+func (cnc *CloudNodeController) getProviderID(ctx context.Context, node *v1.Node) (string, error) {
+	if node.Spec.ProviderID == "" {
+		providerID, err := cloudprovider.GetInstanceProviderID(ctx, cnc.cloud, types.NodeName(node.Name))
+		if err == nil {
+			return providerID, nil
+		} else if err == cloudprovider.NotImplemented {
+			// if the cloud provider being used does not support provider IDs,
+			// we can safely continue since we will attempt to set node
+			// addresses given the node name in getNodeAddressesByProviderIDOrName
+			klog.Warningf("cloud provider does not set node provider ID, using node name to discover node %s", node.Name)
+		} else {
+			// if the cloud provider being used supports provider IDs, we want
+			// to propagate the error so that we re-try in the future; if we
+			// do not, the taint will be removed, and this will not be retried
+			return "", err
+		}
+	}
+
+	return node.Spec.ProviderID, nil
+}
+
+// getInstanceMetadata get instance type and nodeAddresses, use Instances if InstancesV2 is off.
+func (cnc *CloudNodeController) getInstanceMetadata(ctx context.Context, providerID string, nodeName string, needType bool) (*cloudprovider.InstanceMetadata, error) {
```

Comment:
I'm not sure about having `needType` here because when v2 is enabled the type is always returned which can be confusing.

Reply:
How about always getting instance type?

Reply:
> How about always getting instance type?

Done.

Reply:
yeah or maybe a `getNodeAddress` method that only sets `NodeAddress` in `InstanceMetadata`.

Reply:
Not sure yet actually -- let's leave it as is for now (always return instance type in getInstanceMetadata)